fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java

```diff
@@ -0,0 +1,201 @@
+                    .filter(source -> isEquivalent(source, expression))
+                    .map(source -> (Expression) source)
+                    .findFirst();
+            if (found.isPresent()) {
+                return Pair.of(found, true);
+            }
+
+            found = groupByExpressions.stream().filter(source -> isEquivalent(source, expression)).findFirst();
+            return Pair.of(found, false);
+        }
+
+        /**
+         * Check whether the two expressions are equivalent.
+         * @param source The expression in aggregation.
+         * @param expression The expression used to compared to the one in aggregation.
+         * @return true if the expressions are equivalent.
+         */
+        private boolean isEquivalent(Expression source, Expression expression) {
+            if (source.equals(expression)) {
+                return true;
+            } else if (source instanceof Alias) {
+                Alias alias = (Alias) source;
+                return alias.toSlot().equals(expression) || alias.child().equals(expression);
+            }
+            return false;
+        }
+
+        private boolean checkWhetherNestedAggregateFunctionsExist(AggregateFunction function) {
+            return function.children().stream().anyMatch(child -> child.anyMatch(AggregateFunction.class::isInstance));
+        }
+
+        private void generateAliasForNewOutputSlots(Expression expression) {
+            Alias alias = new Alias(expression, expression.toSql());
+            newOutputSlots.add(alias);
+            substitution.put(expression, alias.toSlot());
+        }
+
+        public Map<Expression, Slot> getSubstitution() {
+            return substitution;
+        }
+
+        public List<NamedExpression> getNewOutputSlots() {
+            return newOutputSlots;
+        }
+    }
+
+    private Plan createPlan(LogicalHaving<LogicalAggregate<GroupPlan>> having, Resolver resolver) {
+        LogicalAggregate<GroupPlan> aggregate = having.child();
+        Expression newPredicates = substitute(having.getPredicates(), resolver.getSubstitution());
+        List<NamedExpression> newOutputExpressions = Streams.concat(
+                aggregate.getOutputExpressions().stream(), resolver.getNewOutputSlots().stream()
+        ).collect(Collectors.toList());
+        LogicalFilter<LogicalAggregate<Plan>> filter = new LogicalFilter<>(newPredicates,
+                aggregate.withGroupByAndOutput(aggregate.getGroupByExpressions(), newOutputExpressions));
+        return resolver.getNewOutputSlots().isEmpty()
+                ? filter
+                : new LogicalProject<>(
+                        aggregate.getOutputExpressions().stream().map(NamedExpression::toSlot)
+                                .collect(Collectors.toList()),
+                        filter
+                );
+    }
+
+    private Expression substitute(Expression expression, Map<Expression, Slot> substitution) {
```

Comment:
org.apache.doris.nereids.trees.expressions.visitor.ExpressionReplacer could do same thing.
```java
ExpressionReplacer.INSTANCE.visit(expression, substitution);
```
u could change `ExpressionReplacer` to
```java
extends DefaultExpressionRewriter<Map<? extends Expression, ? extends Expression>>
```
to avoid change substitution map's generic type.

Reply:
Done